### PR TITLE
[example] Domain types rework

### PR DESCRIPTION
**Pull request by TomoLV**
Modify domain types that are used by Scan and PeripheralDetails blocs

### CODE REVIEW

**Review by mikolak — COMMENTED** [review]
Discussed in person

File: example/lib/adapter/ble_adapter.dart [hunk 1]
```diff
@@ -39,37 +49,57 @@ class BleAdapter {
...
-            sink.add(peripheral);
-          }
-        },
-      ),
-    );
+  Stream<ScanResult> startPeripheralScan() {
+    return _bleManager.startPeripheralScan().map((scanResult) {
+      _peripheralContainers.update(
+        scanResult.peripheral.identifier,
+        (_) => PeripheralContainer(
```
Comment (mikolak) [inline example/lib/adapter/ble_adapter.dart:56]: Why not create PeripheralContainer beforehand and only return it in the lambdas?

File: example/lib/adapter/ble_adapter.dart [hunk 2]
```diff
@@ -39,37 +49,57 @@ class BleAdapter {
...
+      return _mapScanResult(scanResult);
+    });
   }
 
   Future<void> stopPeripheralScan() {
     return _bleManager.stopPeripheralScan();
   }
 
+  PeripheralInfo peripheralInfoForIdentifier(String identifier) {
+    PeripheralContainer peripheral = _peripheralContainers[identifier];
```
Comment (mikolak) [inline example/lib/adapter/ble_adapter.dart]: This is not a peripheral.

File: example/lib/peripheral_details/components/peripheral_details_view.dart [hunk 3]
```diff
@@ -1,32 +1,72 @@
...
       ],
     );
   }
+
+  Widget _buildPeripheralIdentifierRow(
+      BuildContext context, PeripheralDetailsBloc peripheralDetailsBloc) {
+    return SliverToBoxAdapter(
+      child: BlocBuilder<PeripheralDetailsBloc, PeripheralDetailsState>(
+        builder: (context, state) {
+          if (state is PeripheralAvailable) {
```
Comment (mikolak) [inline example/lib/peripheral_details/components/peripheral_details_view.dart:33]:
```
if (state is PeripheralAvailable)
  return _buildPeripheralAvailableState();
else if ...
```

File: example/lib/peripheral_details/peripheral_details_bloc.dart [hunk 4]
```diff
@@ -1,22 +1,41 @@
...
+  final String _peripheralIdentifier;
 
-  PeripheralDetailsBloc(this._bleAdapter, this._chosenPeripheral);
+  PeripheralDetailsBloc(this._bleAdapter, this._peripheralIdentifier);
 
   @override
-  PeripheralDetailsState get initialState =>
-      PeripheralDetailsState(peripheral: _chosenPeripheral);
+  PeripheralDetailsState get initialState {
+    return _refreshStateFromAdapter();
```
Comment (mikolak) [inline example/lib/peripheral_details/peripheral_details_bloc.dart:16]: I'm not a fan of this. Perhaps BLoC should get name as well as a parameter and display only those two in the initial state and fetch any additional info later?

File: example/lib/peripheral_details/peripheral_details_screen.dart [hunk 5]
```diff
@@ -9,48 +9,72 @@ class PeripheralDetailsScreen extends StatelessWidget {
...
-            ),
-            body: _buildDetailsView(),
-          );
+          return null;
         }
       },
     );
   }
 
+  Widget _buildForPeripheralAvailable(PeripheralAvailable state) {
```
Comment (mikolak) [inline example/lib/peripheral_details/peripheral_details_screen.dart]: `for`? What's it for?

File: example/lib/peripheral_details/peripheral_details_screen.dart [hunk 6]
```diff
@@ -9,48 +9,72 @@ class PeripheralDetailsScreen extends StatelessWidget {
...
+    if (state.peripheralInfo.peripheralLayout ==
```
Comment (mikolak) [inline example/lib/peripheral_details/peripheral_details_screen.dart:24]: If there's an if with two long branches, you'd be better off having those branches in separate functions. Makes for an easier read.

File: example/lib/scan/components/scan_result_item.dart [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    return PropertyRow(
+      title: _scanResult.identifier,
+      titleIcon: AssetManager.iconForPeripheral(context, _scanResult.category),
+      titleColor: ColorManager.colorForPeripheral(context, _scanResult.category),
+      value: _scanResult.name,
+      titleAccessory: Icon(
+        Icons.chevron_right,
+        color: Colors.grey,
+      ),
+      valueAccessory: Row(
```
Comment (mikolak) [inline example/lib/scan/components/scan_result_item.dart:28]: `_buildValueAccessory()`?

File: example/lib/scan/scan_bloc.dart [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+class ScanBloc extends Bloc<ScanEvent, ScanState> {
+  BleAdapter _bleAdapter;
+  StreamSubscription _scanResultsSubscription;
+  Map<String, ScanResult> _scanResults = {};
+
+  ScanBloc(this._bleAdapter);
+
+  @override
+  ScanState get initialState => ScanState.initial();
```
Comment (mikolak) [inline example/lib/scan/scan_bloc.dart:16]: `ScanState.initial()`? I assume it initialises all values to some safe defaults, but then why not just make those optional in normal constructor?

File: example/lib/scan/scan_bloc.dart [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+    _scanResults.update(
+      event.scanResult.identifier,
+      (_) => event.scanResult,
+      ifAbsent: () => event.scanResult,
+    );
+
+    yield ScanState(
+        scanResults: _scanResults.values
+            .map((scanResult) => scanResult.viewModel())
+            .toList(),
```
Comment (mikolak) [inline example/lib/scan/scan_bloc.dart:56]: `_toScanItems(List<ScanResult>)` or something like that.

File: example/lib/scan/scan_result_view_model.dart [hunk 10]
```diff
@@ -0,0 +1,37 @@
+import 'package:blemulator_example/scan/scan_result.dart';
+import 'package:equatable/equatable.dart';
+
+class ScanResultViewModel extends Equatable {
```
Comment (mikolak) [inline example/lib/scan/scan_result_view_model.dart:4]: `ScannedPeripheral`? `ScanItem`? There are a lot of namespace conflicts after all.

File: example/lib/scan/scan_state.dart [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+import 'package:flutter/foundation.dart';
+
+class ScanState extends Equatable {
+  final List<ScanResultViewModel> scanResults;
+  final bool scanningEnabled;
+
+  const ScanState({@required this.scanResults, @required this.scanningEnabled});
+
+  const ScanState.initial(
+      {this.scanResults = const [], this.scanningEnabled = false});
```
Comment (mikolak) [inline example/lib/scan/scan_state.dart:12]: This is essentially the same as default constructor without `@required`

File: example/lib/styles/custom_colors.dart [hunk 12]
```diff
@@ -8,4 +8,6 @@ abstract class CustomColors {
 
   static const rowCardBackground = Colors.white;
   static const scaffoldBackground = Color.fromRGBO(242, 242, 247, 1.0);
+
+  static const sensorTag = Color(0xFFCC0000);
```
Comment (mikolak) [inline example/lib/styles/custom_colors.dart]: `sensorTagRed`?

File: example/lib/util/asset_manager.dart [hunk 13]
```diff
@@ -0,0 +1,32 @@
+import 'package:blemulator_example/common/components/title_icon.dart';
+import 'package:blemulator_example/common/components/title_image_icon.dart';
+import 'package:blemulator_example/scan/scan_result.dart';
+import 'package:blemulator_example/scan/scan_result_view_model.dart';
+import 'package:blemulator_example/util/color_manager.dart';
+import 'package:flutter/material.dart';
+import 'package:flutter/widgets.dart';
+
+class AssetManager {
```
Comment (mikolak) [inline example/lib/util/asset_manager.dart]: `IconManager`?

**Review by mikolak — CHANGES_REQUESTED** [review]
Still got some concerns. We can pair tomorrow, if you'd like.

I'd also appreciate it if the automated tests of example were moved to a separate CI stage.

File: example/lib/adapter/ble_adapter.dart (hunk 2 above)
Comment (TomoLV) [reply example/lib/adapter/ble_adapter.dart]: Renamed to `peripheralContainer`

File: example/lib/peripheral_details/components/peripheral_details_view.dart (hunk 3 above)
Comment (TomoLV) [reply example/lib/peripheral_details/components/peripheral_details_view.dart:33]: Extracted to a separate method

File: example/lib/peripheral_details/peripheral_details_screen.dart (hunk 5 above)
Comment (TomoLV) [reply example/lib/peripheral_details/peripheral_details_screen.dart]: Renamed to `_buildPeripheralAvailableState`

File: example/lib/util/asset_manager.dart (hunk 13 above)
Comment (TomoLV) [reply example/lib/util/asset_manager.dart]: Renamed to `IconManager`

File: example/lib/styles/custom_colors.dart (hunk 12 above)
Comment (TomoLV) [reply example/lib/styles/custom_colors.dart]: Renamed to `sensorTagRed`

**Pull request closed without merging**